File: revisions/plot_LD_single_rep.R
library(dplyr)
library(tidyr)
library(DBI)
library(ggplot2)
library(tibble)
library(viridis)

df = tibble()
rhovals = c(1e2,1e3,1e4)
I=1
twoN = 1e4
for(directory in c("lowrec", "main_results", "hirec"))
{
    db = dbConnect(RSQLite::SQLite(),paste(directory,"/himu/replicate1_ld.sqlite3",sep=""))
    dbt = tbl(db, 'data')
    for (minc2 in c(1,0.05*twoN, 0.25*twoN))
    {
        q = dbt %>% select(generation,pos1,pos2,e1,e2,c1,c2,D) %>%
            filter(generation <= 60000, c2 >= minc2, c2 <= twoN - minc2, 0.5*twoN*e1*e1 >= 10) %>%
            mutate(intralocus = ifelse(abs(pos1-pos2) < 1.0, 1, 0),
                   positive = ifelse(e1 > 0, '+', '-'),
                   positive2 = ifelse(e2 > 0, '+', '-')) %>%
            group_by(generation, pos1, positive, positive2, intralocus) %>% summarise(mD=mean(D),mine=min(e2),maxe=max(e2))
            #group_by(generation, pos1, positive) %>% summarise(mD=mean(D))
            #group_by(generation, pos1) %>% summarise(mD=mean(D))
        t = collect(q) %>% mutate(rho = rhovals[I],minc2 = minc2/twoN)
        #print(head(t))
        # x = t %>% group_by(pos1, positive, positive2, intralocus) %>% summarise(n=n())
        # print(x)
        df = bind_rows(df, t)
    }
    dbDisconnect(db)
    I <- I + 1
}

label_rho <- function(x)
{
    return (expression(paste(rho," = ",x)))
}

label_count <- function(x)
{
    return (paste("MAF =",as.numeric(x)/twoN,sep=' '))
}

df = df %>% filter(intralocus == 1)
p = ggplot(df,aes(x=generation-5e4,y=mD,color=as.factor(pos1),shape=positive2)) + geom_point(alpha=0.25,size=0.05) + 
    scale_color_viridis(discrete=TRUE) +
    #facet_grid(minc2~as.factor(rho),labeller=labeller(minc2=label_count,rho=as_labeller(label_rho))) + 
    facet_grid(minc2*positive2~rho*positive,
               labeller=label_bquote(cols=rho ~"="~ .(rho)~","~.(positive),rows="MAF >="~.(minc2) ~", "~.(positive2))) +
    theme_bw(base_size=14) +
    theme(legend.position='none',axis.text.x=element_text(angle=45,vjust=1,hjust=1)) + 
    xlab("Generations since optimum shift") + 
    ylab("Mean disequilibrium statistic, D")

ggsave("LDsinglerep.pdf",height=10,width=10,units="in")

# up = unique(df$pos1)
# 
# t = df %>% filter(pos1==up[1])
# plot(t$generation, t$mD)
# for (i in up[2:length(up)])
# {
#     t = df %>% filter(pos1==i)
#     lines(t$generation, t$mD)
# }
# 

File: revisions/plot_LD_single_rep.py
import sqlite3
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import pandas as pd
import numpy as np

THRESHOLD = np.sqrt(50. / 5e3)

fig = plt.figure()
gs = gridspec.GridSpec(3, 3)
freqaxes = []
ldaxes = []

freqaxes.append(fig.add_subplot(gs[0, 0]))
freqaxes.append(fig.add_subplot(gs[0, 1],sharex=freqaxes[0]))
freqaxes.append(fig.add_subplot(gs[0, 2],sharex=freqaxes[0]))

ldaxes.append(fig.add_subplot(gs[1, 0], sharex=freqaxes[0]))
ldaxes.append(fig.add_subplot(gs[2, 0], sharex=freqaxes[0], sharey=ldaxes[0]))

ldaxes.append(fig.add_subplot(gs[1, 1], sharex=freqaxes[1], sharey=ldaxes[0]))
ldaxes.append(fig.add_subplot(gs[2, 1], sharex=freqaxes[1], sharey=ldaxes[0]))

ldaxes.append(fig.add_subplot(gs[1, 2], sharex=freqaxes[2], sharey=ldaxes[0]))
ldaxes.append(fig.add_subplot(gs[2, 2], sharex=freqaxes[2], sharey=ldaxes[0]))

directories = ['lowrec', 'main_results', 'hirec']
replicates=[199]*3
replicates=[189]*3
replicates=[169]*3
replicates=[1]*3
for i, d in enumerate(directories):
    conn = sqlite3.connect('{}/himu/replicate{}_ld.sqlite3'.format(d,replicates[i]))
    df = pd.read_sql(
        'select * from data where abs(pos2-pos1) < 1.0 and generation < 60000 and abs(e1) >= {}'.format(THRESHOLD), conn)

    # plot freqs in first plot
    groups = df.groupby(['pos1'])
    for n, g in groups:
        print(g.e1.unique())
        freqaxes[i].plot(g.generation - 5e4, g.c1 / 1e4,
                         label='{:0.3f}'.format(g.e1.unique()[0]))
        freqaxes[i].legend(loc='lower right', fontsize='small', edgecolor=None,
                framealpha=0.0)

    dplotneg = ldaxes[2 * i]
    dplotpos = ldaxes[2 * i + 1]
    minc2 = 0.10 * 1e4
    alpha = 1
    for n, g in groups:
        g2 = g[(g.c2 >= minc2) & (g.c2 < 1e4 - minc2) & (g.e2 < 0)]
        g2 = g2.groupby(['generation']).mean().reset_index()
        dplotneg.scatter(g2.generation - 5e4, g2.D, s=0.05, alpha=alpha)
        g2 = g[(g.c2 >= minc2) & (g.c2 < 1e4 - minc2) & (g.e2 > 0)]
        g2 = g2.groupby(['generation']).mean().reset_index()
        dplotpos.scatter(g2.generation - 5e4, g2.D, s=0.05, alpha=alpha)

# hide axes
for i in freqaxes:
    i.axes.get_xaxis().set_visible(False)
for i in freqaxes[1:]:
    i.axes.get_yaxis().set_visible(False)
for i in ldaxes[0:5:2]:
    i.axes.get_xaxis().set_visible(False)
for i in ldaxes[2:]:
    i.axes.get_yaxis().set_visible(False)

# axis labels
ldaxes[3].set_xlabel("Generations since optimum shift")
freqaxes[0].set_ylabel("Frequency")
ldaxes[0].set_ylabel(r'$\bar{D}, \gamma < 0$')
ldaxes[1].set_ylabel(r'$\bar{D}, \gamma > 0$')

# column titles
for i,j in zip(freqaxes, [1e2,1e3,1e4]):
    i.set_title(r'$\rho = {}$'.format(j))

fig.align_ylabels()
plt.savefig('LDsinglerep.pdf')
